experimental/virtio/src/vsock/socket.rs

```diff
@@ -0,0 +1,364 @@
+            packet.set_flags(VSockFlags::all());
+            self.config.vsock.write_packet(&mut packet);
+        }
+    }
+
+    /// Whether we should send an unsolicited credit update.
+    fn must_send_credit_update(&self) -> bool {
+        STREAM_BUFFER_LENGTH - (self.processed_bytes - self.previous_processed_bytes)
+            < CREDIT_UPDATE_LIMIT
+    }
+
+    /// Sends a control packet with the specified op to the host.
+    fn send_control_packet(&mut self, op: VSockOp) -> anyhow::Result<()> {
+        // For now we panic if we are disconnected.
+        assert!(self.connection_state == ConnectionState::Connected);
+        let mut packet = Packet::new_control(self.config.port, self.config.host_port, op)?;
+        self.set_credit_info(&mut packet);
+        self.config.vsock.write_packet(&mut packet);
+        Ok(())
+    }
+
+    /// Sends a data packet to the host.
+    fn send_data_packet(&mut self, data: &[u8]) -> anyhow::Result<()> {
+        // For now we panic if we are disconnected.
+        assert!(self.connection_state == ConnectionState::Connected);
+        let data_len = data.len();
+        assert!(
+            data_len <= MAX_PAYLOAD_SIZE,
+            "The data is too large for a single packet. Len: {}, Max: {}",
+            data.len(),
+            MAX_PAYLOAD_SIZE
+        );
+
+        let data_len = Wrapping(data_len as u32);
+        if data_len > self.peer_buffer_size - (self.sent_bytes - self.peer_processed_bytes) {
+            anyhow::bail!("Peer's stream buffer is full.");
+        }
+
+        self.sent_bytes += data_len;
+        let mut packet = Packet::new_data(data, self.config.port, self.config.host_port)?;
+        self.set_credit_info(&mut packet);
+        self.config.vsock.write_packet(&mut packet);
+        Ok(())
+    }
+
+    /// Updates the credit info on a packet to facilitate flow-control.
+    fn set_credit_info(&mut self, packet: &mut Packet) {
+        packet.set_buf_alloc(STREAM_BUFFER_LENGTH.0);
+        packet.set_fwd_cnt(self.processed_bytes.0);
+        self.previous_processed_bytes = self.processed_bytes;
+    }
+
+    /// Reads the payload of the next available data packet, if any are available.
+    fn read_data(&mut self) -> Option<VecDeque<u8>> {
+        // For now we panic if we are disconnected.
+        assert!(self.connection_state == ConnectionState::Connected);
+        let src_port = self.config.host_port;
+        let dst_port = self.config.port;
+        loop {
+            let packet = self.config.vsock.read_filtered_packet(
+                |packet| packet.get_dst_port() == dst_port && packet.get_src_port() == src_port,
+                true,
+            )?;
+            self.peer_buffer_size = Wrapping(packet.get_buf_alloc());
+            self.peer_processed_bytes = Wrapping(packet.get_fwd_cnt());
+            // For now we panic if we receive an invalid op.
+            match packet.get_op().expect("Invalid packet received on stream.") {
```

Comment:
Could `unwrap`s in this block be replaced with `expect`?

Reply:
Done.